src/engine/SCons/Node/FS.py

```diff
@@ -3243,9 +3262,127 @@ def changed_content(self, target, prev_ni):
     def changed_state(self, target, prev_ni):
         return self.state != SCons.Node.up_to_date
 
-    def changed_timestamp_then_content(self, target, prev_ni):
+
+    def _build_dependency_map(self, binfo):
+        """
+        Build mapping from file -> signature
+
+        This method also updates binfo.dependency_map to avoid rebuilding
+        the map for every dependency of the node we're workingon.
+
+        Args:
+            self - self
+            binfo - buildinfo from node being considered
+
+        Returns:
+            dictionary of file->signature mappings
+        """
+
+        # For an "empty" binfo properties like bsources
+        # do not exist: check this to avoid exception.
+        if (len(binfo.bsourcesigs) + len(binfo.bdependsigs) + \
+                len(binfo.bimplicitsigs)) == 0:
+            return {}
+
+        pairs = [
+            (binfo.bsources, binfo.bsourcesigs),
```

Comment:
My SCons-fu has gotten a bit rusty over the course of the last year. This approach assumes that the lists 'children' and 'signatures' always have the same length, correct? Or do you let "zip" fix things for you if required?